### PR TITLE
Home work24.1

### PR DESCRIPTION
**Pull request by KostiantynVorobiov**

### CODE REVIEW

File: src/main/java/com/internet/shop/models/Product.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package com.internet.shop.models;
+
+public class Product {
+    private Long id;
+    private String name;
+    private Double price;
```
Comment (vremere-a) [inline src/main/java/com/internet/shop/models/Product.java]: for advise - use BigDecimal.

**Review by vremere-a — APPROVED** [review]

File: src/main/java/com/internet/shop/Application.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import com.internet.shop.service.ProductService;
+
+public class Application {
+    private static Injector injector = Injector.getInstance("com.internet.shop");
+
+    public static void main(String[] args) {
+        ProductService productService = (ProductService) injector.getInstance(ProductService.class);
+
+        Product productForUpdate = new Product("Motorola", 50.50);
+        Product fistProduct = new Product("Samsung", 7000.55);
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/Application.java]: let's use more informative names like `motorola`, `samsung`

File: src/main/java/com/internet/shop/Application.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        productService.create(new Product("Xiaomi", 3245.10));
+        productService.create(new Product("Nokia", 6215.99));
+        productService.create(productForUpdate);
+        System.out.println(productService.getAll());
+        for (Product product : productService.getAll()) {
+            System.out.println(product);
+        }
+        System.out.println();
+
+        System.out.println("Read product by ID " + productService.get(3L));
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/Application.java]: save your products to variables and use something like `huawei.getId()` instead where needed

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Override
+    public Optional<Product> getById(Long id) {
+        for (int i = 0; i < Storage.products.size(); i++) {
+            if (Storage.products.get(i).getId() == id) {
+                return Optional.ofNullable(Storage.products.get(i));
+            }
+        }
+        return Optional.empty();
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: let's use Stream API and `findFirst()`

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        return Storage.products;
+    }
+
+    @Override
+    public Product update(Product product) {
+        for (int i = 0; i < Storage.products.size(); i++) {
+            if (Storage.products.get(i).getId().equals(product.getId())) {
+                return Storage.products.set(i, product);
+            }
+        }
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: the logic is ok, but let's use Stream API

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Override
+    public boolean deleteById(Long id) {
+        for (int i = 0; i < Storage.products.size(); i++) {
+            if (Storage.products.get(i).getId() == id) {
+                Storage.products.remove(i);
+            }
+        }
+        return true;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: you can use `removeIf()` on storage

File: src/main/java/com/internet/shop/models/Order.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+import java.util.List;
+
+public class Order {
+    private Long id;
+    private List<Product> products;
+    private Long userId;
+
+    public Order(List<Product> products, Long userId) {
+        this.products = products;
+        this.userId = userId;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/models/Order.java]: we will not receive `products` from outside

File: src/main/java/com/internet/shop/models/Product.java (hunk 1 above)
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/models/Product.java]:
```suggestion
    private double price;
```
using a wrapper is recommended for id, but for this field we can use a primitive type

File: src/main/java/com/internet/shop/models/ShoppingCart.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+import java.util.List;
+
+public class ShoppingCart {
+    private Long id;
+    private List<Product> products;
+    private Long userId;
+
+    public ShoppingCart(List<Product> products, Long userId) {
+        this.products = products;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/models/ShoppingCart.java]: same: we will not receive products from outside

File: src/main/java/com/internet/shop/models/User.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
+package com.internet.shop.models;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/models/User.java]:
```suggestion
package com.internet.shop.model;
```

let's name packages in singular

File: src/main/java/com/internet/shop/service/impl/ProductServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    private ProductDao productDao;
+
+    @Override
+    public Product create(Product product) {
+        return productDao.create(product);
+    }
+
+    @Override
+    public Product get(Long id) {
+        return productDao.getById(id).orElseThrow();
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/service/impl/ProductServiceImpl.java]:
```suggestion
        return productDao.getById(id).get();
```
let's use `get()` for now

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
also, remove `<dependencies>` block from pom.xml
we do not use dependencies yet

File: src/main/java/com/internet/shop/Application.java (hunk 2 above)
Comment (KostiantynVorobiov) [reply src/main/java/com/internet/shop/Application.java]: Done

File: src/main/java/com/internet/shop/Application.java (hunk 3 above)
Comment (KostiantynVorobiov) [reply src/main/java/com/internet/shop/Application.java]: Done

File: src/main/java/com/internet/shop/models/Order.java (hunk 7 above)
Comment (KostiantynVorobiov) [reply src/main/java/com/internet/shop/models/Order.java]: Done

File: src/main/java/com/internet/shop/models/Product.java (hunk 1 above)
Comment (KostiantynVorobiov) [reply src/main/java/com/internet/shop/models/Product.java]: Done

File: src/main/java/com/internet/shop/models/User.java (hunk 9 above)
Comment (KostiantynVorobiov) [reply src/main/java/com/internet/shop/models/User.java]: Done

File: src/main/java/com/internet/shop/models/ShoppingCart.java (hunk 8 above)
Comment (KostiantynVorobiov) [reply src/main/java/com/internet/shop/models/ShoppingCart.java]: Done

File: src/main/java/com/internet/shop/service/impl/ProductServiceImpl.java (hunk 10 above)
Comment (KostiantynVorobiov) [reply src/main/java/com/internet/shop/service/impl/ProductServiceImpl.java]: Done

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java (hunk 6 above)
Comment (KostiantynVorobiov) [reply src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: Done

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java (hunk 4 above)
Comment (KostiantynVorobiov) [reply src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: Done

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java (hunk 5 above)
Comment (KostiantynVorobiov) [reply src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: Done

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    @Override
+    public Product create(Product product) {
+        Storage.addProduct(product);
+        return product;
+    }
+
+    @Override
+    public Optional<Product> getById(Long id) {
+        return Storage.products.stream()
+                .filter(product -> product.getId() == id)
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: how do we compare object types?

File: src/main/java/com/internet/shop/model/Order.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Order {
+    private Long id;
+    private List<Product> products;
+    private Long userId;
+
+    public Order(List<Product> products, Long userId) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/Order.java]:
```suggestion
    public Order(Long userId) {
```

File: src/main/java/com/internet/shop/model/ShoppingCart.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ShoppingCart {
+    private Long id;
+    private List<Product> products;
+    private Long userId;
+
+    public ShoppingCart(List<Product> products, Long userId) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/ShoppingCart.java]: same here

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**